### PR TITLE
fix: name resolution should respect package/namespace

### PR DESCRIPTION
**Pull request by barnasc**
**Environment**

- `sysml-v2-lsp` version `0.27.0` only (No `VSCode_SysML_Extension` used)

**Problem**

`checkUnresolvedType` (`server/src/providers/semanticValidator.ts`) flags a type reference as unresolved unless its bare name matches *any* symbol in the entire workspace symbol table (`allSymbolNames = new Set(symbolTable.getAllSymbols().map(s => s.name))`), or the standard library. It never checks whether that symbol is actually visible from the reference site: same
namespace, an enclosing namespace, or an explicit `import`.

In practice this means two unrelated packages that happen to reuse the same short name (e.g. both defining `Part3`) silently resolve against each other with no `import` connecting them, masking a genuine unresolved-reference error. The diagnostic's own message ("Type 'X' is not defined in the current document or standard library") also misleadingly implies document-local
resolution, when the check is actually workspace-wide.

**Reproducer:**

```sysml
// pkg-a.sysml
package PkgA { part def Part3; }

// pkg-b.sysml -- no import of PkgA
package PkgB {
    part def Part1;
    part usesPart3 : Part3; // should be unresolved-type, but isn't
}
```

**Fix**

Per the SysML v2 spec §7.5.1/§7.5.3, an `import` doesn't grant one-off visibility at reference time: it creates an actual (unowned) membership in the importing namespace, on equal footing with a directly-declared member, which is why imported members are themselves further importable (transitively, if the import was `public`). This fix implements this:

- Name resolution now respects package/namespace membership and `import` statements: a type reference only resolves if it's actually visible from the reference site, declared in the same or an enclosing namespace, or reachable through an `import`; rather than matching any symbol with that name anywhere in the workspace.
- Imports are modeled as real memberships of the importing namespace, not a one-off visibility check, so an imported name is itself further importable by whoever imports *that* namespace, transitively, if the import was `public`; not at all, if `private` or `protected`. All four import forms are covered: importing one specific name, importing a namespace's direct members, and the recursive (`::**`) form of each.
- §7.5.4 import filtering is supported: a package-level `filter` condition (applies to every import of that package) and per-import filtered imports (a condition attached to one specific import). Filter conditions that test for the presence of metadata are evaluated; conditions this doesn't understand (e.g. comparisons against an attribute's value) are treated as passing rather
 than silently blocking an otherwise-valid import; see *known-gaps* note below for what that leaves unenforced.
- Qualified references (`Owner::Name`) now verify that `Name` is an actual member of `Owner`, instead of only checking that `Owner` is a known name somewhere in the workspace and assuming whatever follows it is fine. A qualified reference to a real member of an unimported top-level package still correctly resolves either way, per §7.5.5 (top-level elements are name-resolvable from anywhere, no import required). This only tightens the check on what comes after `::`.
- The same fix also applies to the `sysml/model` request's own diagnostics (Model Explorer, Dashboard, Feature Inspector, status bar), not just the editor's `textDocument/publishDiagnostics` output.

**Testing**

Added 20 test cases to `test/unit/diagnostics.test.ts` and 2 to `test/unit/sysmlModel.test.ts`:

- An unimported cross-package reference is flagged as unresolved.
- Exact (`import Pkg::Name;`), shallow (`import Pkg::*;`), and deep (`import Pkg::**;`) imports all resolve correctly.
- An import of an unrelated package does not grant access to an unimported name.
- Same-package definitions remain visible across files without any import.
- Resolution works for non-`part` definition kinds (`attribute def`, `port def`, `interface def`) and for an imported usage, not just definitions.
- The standard's own `P2`/`Q` re-import and `P4`/`P5`/`P6` recursive-import examples are reproduced verbatim.
- A publicly imported name propagates to a further importer; a privately or protected-imported name does not.
- A package-level filter, an inline filtered import, `and`/`not` filter composition, and an unsupported filter expression (fails open) are each covered.
- A qualified reference to a real member of an unimported top-level package still resolves correctly; a qualified reference to a name that is *not* actually a member of the qualifying package is correctly flagged (the regression case).
- The `sysml/model` request's own diagnostics (not just the editor's) flag an unimported cross-package reference and do not flag a type declared in a different document of the same workspace.

Full existing suite passes unchanged (no regressions).

**Known limitation**

- Filter expressions support `@Name` metadata-presence tests combined with `and`/`or`/`not` and simple parenthesization. The grammar's `ALL typeReference` expression form (a filter condition meaning "all instances of a type") is not evaluated. It isn't recognized as a distinct case, so it's treated the same as any other unsupported expression (e.g. an attribute-value comparison)
  and evaluates as passing rather than actually filtering. This fails open, so it doesn't hide an otherwise-valid import, but it also means a filter written this way isn't enforced.

*Developed and tested by @barnasc. Assisted-by: Claude Sonnet 5*

### CODE REVIEW

**Comment by daltskin** [comment]
@barnasc many thanks for tackling this.  Please can you address the following before merge:

- Support packages split across files. Imports from every fragment of the same package must be combined. Currently, only one fragment’s imports are retained, causing valid references in other files to be reported as unresolved.
- 
- Resolve relative imports correctly. An import such as import C; inside a nested package must resolve C from its enclosing namespaces. The current lookup treats it as an absolute name, so re-exporting it to another package fails.

Please add regression tests covering both cases, including an external package consuming the relative re-export. Also add an entry under CHANGELOG.md → Unreleased.

The existing test suite, compile, and lint checks otherwise pass.

**Comment by barnasc** [comment]
THanks for your review comment and guidance! 

The update contains:
- Fixed: a SysML package split across multiple files now correctly combines the imports declared in each file, instead of only picking up whichever file happened to be processed last.
- Fixed: a name re-imported inside a nested package (relying on an outer package's own import) now resolves correctly, and can itself be passed on to a further package.
- Fixed: a privately or protected imported name is no longer incorrectly reachable from outside the package that imported it, when accessed via a fully-qualified path.
- Fixed: if the same name is imported with different visibility (e.g. private in one file, public in another file of the same package), the result no longer depends on which file happens to be processed first — the more permissive visibility now always wins (i.e. public wins from protected and private).
- Added 18 new regression tests covering all of the above, in both directions (should-resolve and should-not-resolve cases).
- Updated CHANGELOG.md

**Comment by daltskin** [comment]
Thanks for addressing the earlier feedback and adding the regression tests and changelog entry. The full suite and compilation pass, and lint reports no errors. Additional testing found three remaining resolution issues to address before merging:

Removed imports remain visible across package fragments. Keep each fragment’s original imports and filters separate from the merged package representation. Rebuild the aggregate from the remaining fragments after edits or removal, without writing merged data back into a source fragment. Add an incremental test that removes an import and verifies the reference becomes unresolved.

Relative imports skip the importing namespace itself. Begin lookup in the namespace containing the import, then walk outward. Ensure cycle handling exposes already-known owned members during resolution, rather than an empty placeholder. Test importing a directly nested package and consuming its public re-export from an external package.

Root names take precedence over local names. Search from the innermost namespace outward, with the implicit root last. Once the first segment resolves locally, resolve subsequent segments against that namespace without falling back to a same-named outer package. Add a regression where local and root packages share a name but contain different members.

**Comment by barnasc** [comment]
thanks for listing these additional corner/test cases. While reviewing the SysMLv2 standard, I found a few other cases which are now also addressed in the PR update:


- Removed imports no longer stay visible: fixed by building a fresh, non-mutating merged copy each time instead of writing the merge back into a fragment's own object.
- Relative import target resolution now begins in the importing namespace itself, not its parent: fixes importing a package declared directly inside the same namespace as the import statement (previously broken due to an overly-cautious cycle guard; now the guard seeds with already-known owned members instead of an empty table).
- Local names now correctly shadow same-named root-level names: name resolution was checking the root namespace before the innermost enclosing scope (backwards); fixed to search innermost-outward with root last, per §7.5.1.
- Imports/filters inside a definition or usage body are now extracted: previously only package bodies were scanned, silently discarding any import written inside a part definition.
- Protected import visibility's specialization exception (§7.5.3): a specialization of a definition/usage can now see its protected-imported members via a qualified reference; previously protected behaved identically to private everywhere, with no specialization awareness at all.
- Package fragment merge completeness — documentation, metadataAnnotations, and viewFilters are now combined across a split package's fragments too, not just imports/filters; a doc-comment or metadata annotation on a non-canonical fragment was previously silently dropped.
- Specialization check no longer matches unrelated same-named definitions: resolves it the same namespace-aware way an ordinary reference would.

Known limitations / out-of-scope for this PR, to be addressed separately:

- Root-level import, owned directly by the implicit root namespace, outside any package (see §7.5.3 constraint, root imports must be forced private).
- Filter body-metadata form (`@Name { ... }`) and attribute-value comparisons.
- Unqualified (non-qualified-path) inherited access to a protected member via specialization. Requires full §7.6 feature-inheritance modeling, deliberately out of scope for this resolver.

**Comment by daltskin** [comment]
Thanks for the updates and additional coverage.  The earlier fixes look good, and the full test suite and compilation pass.

Testing the expanded changes uncovered two issues we should address before merging:

- Recursive imports through mutually re-exported packages can cause a stack overflow. Please guard against cycles during traversal.
- A nested package’s imports can be attributed to its enclosing definition, replacing the definition’s own imports. Please restrict extraction to the owning namespace’s body.

Please add focused regression tests for these two cases. We also found that imports depending on another import in the same namespace can remain unresolved, but we’re comfortable tracking that separately.

Appreciate the work on this!

**Comment by barnasc** [comment]
Thanks for reviewing and listing additional test/corner cases. Just out of curiosity, how did you find these?
Resolution of the items:

- Stack overflow guard: Verified and was already safe. The existing cache-seeding mechanism breaks any import cycle, confirmed by removing it and reproducing a real stack overflow. No fix needed, but added 3 regression tests (mutual, self-referential, 3-node cycles).

- Nested package imports leaking into enclosing definition: Fixed, import extraction searched for a package body unconditionally with no boundary, so a nested package's import silently replaced its enclosing definition's own. Restricted extraction to the owning namespace's own body, with a regression test reproducing the exact leak.

I also fixed the same-namespace import ordering, since it was related to the potential stack overflow item. While fixing the cache mechanism, the risk for a stack overflow increased, so I had to verify/fix both in lock-step. Added regression test for this case as well.

While investigating and creating more regression tests, I found another interesting corner case: two declarations sharing the same name, e.g. `package A` and `part def A`. According to the KerML spec (nothing is said in the SysML spec), this seems a valid case: two elements only conflict if their kinds conform to each other, not merely their names. Therefore tests are added with same-kind collisions (e.g. two `part def A`), and added a diagnostic (`ambiguous-namespace-name`) for that case, and rewrote the regression tests to match this rule.

I welcome your feedback and any additional input/guidance you might have.

File: server/src/symbols/namespaceResolver.ts [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+    private isSpecializationOf(candidateQualifiedName: string, ownerQualifiedName: string, indexes: SymbolIndexes): boolean {
+        const visited = new Set<string>([candidateQualifiedName]);
+        let frontier = [candidateQualifiedName];
+        let guard = 0;
+        while (frontier.length > 0 && guard++ < 64) {
+            const next: string[] = [];
+            for (const qualifiedName of frontier) {
+                const symbol = indexes.byQualifiedName.get(qualifiedName);
+                for (const typeName of symbol?.typeNames ?? []) {
+                    const supertype = this.resolveQualifiedNameFrom(symbol?.parentQualifiedName, typeName, indexes);
```
Comment (daltskin) [inline server/src/symbols/namespaceResolver.ts]: **Blocking:** resolving a supertype here can re-enter protected-member visibility (`resolveQualifiedNameFrom` -> `isProtectedVisibleFrom` -> `isSpecializationOf` -> this call) without sharing a recursion guard, causing `SemanticValidator.validate()` to throw `RangeError: Maximum call stack size exceeded` for valid SysML syntax.

Reproducer:
```sysml
package Lib { part def Engine; }
part def Vehicle { protected import Lib::Engine; }
part def Container :> Container::Child {
    part def Child :> Vehicle::Engine;
    part engine : Vehicle::Engine;
}
```

Please carry a re-entrancy/cycle guard across this call chain and add a regression test asserting validation returns diagnostics instead of throwing. The existing `guard < 64` only bounds the local loop; each recursive call starts a new guard.

**Review by daltskin — CHANGES_REQUESTED** [review]
Thanks for the substantial follow-up work. The earlier import-cycle, package-fragment, relative-import, shadowing, and nested-import issues now look addressed, and the full local suite passes. One stack-overflow path remains in protected-import specialization resolution; I’ve left a focused inline reproducer. Please add a shared re-entrancy guard and regression test before merge.

File: server/src/symbols/namespaceResolver.ts (hunk 1 above)
Comment (barnasc) [reply server/src/symbols/namespaceResolver.ts]: Thanks for catching this. Addressed in [2daa86f](https://github.com/daltskin/sysml-v2-lsp/pull/105/commits/2daa86fc684ce3daa949dbddb631840d6559a3b7)

**Comment by barnasc** [comment]
Updates in [2daa86f](https://github.com/daltskin/sysml-v2-lsp/pull/105/commits/2daa86fc684ce3daa949dbddb631840d6559a3b7):

- fixed incorrect visibility of protected member when in a specialization chain, by introducing a re-entrance guard. Added regression test
- `or` filter expression test added (`or` check was missing from regression)
- incremental regression test added: changing a conflicting package declaration wasn't cleaned up correctly and stuck around permanently.
- rephrased two older inline comments describing the old pre-correction conflict rule — updated for consistency with latest code base
- incremental visibility tests (e.g. public → private/protected and reverted, changed metadata annotation and filter) to check next validation. Working as expected.
- general code cleanup and readability improvements
- updated changelog explaining all relevant updates

**Review by daltskin — APPROVED** [review]
Re-reviewed at `2daa86f`. The protected-visibility specialization recursion is now guarded with scoped `try/finally` cleanup, and the original valid-syntax reproducer returns the expected unresolved-type diagnostics instead of throwing. The new incremental conflict/visibility/filter coverage also looks sound. Local full suite: 449 tests passed; compilation passed; lint has zero errors (one existing generated-code warning). Approved.

**Comment by daltskin** [comment]
Many thanks @barnasc! 

For transparency I'm using both manual and copilot to help with maintenance & reviews.
